### PR TITLE
WIP - pkg/flock package

### PR DESCRIPTION
**Pull request by vrothberg**
Introduce the pkg/flock package, a flock(2)-based implementation in
golang that can be used within and across process spaces. The main
motivation for this package is to benefit from the TryLock-semantics
offered by flock(2), which is not possible with c/storage.Locker file
lock since they are based on fcntl(2).

Signed-off-by: Valentin Rothberg <rothberg@redhat.com>

### CODE REVIEW

File: pkg/flock/flock_windows.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
+// build unix
```
Comment (giuseppe) [inline pkg/flock/flock_windows.go]: should this be dropped?

File: pkg/flock/flock_unix.go [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+	return flock{
+		flockType: flockUnlocked,
+		fd:        -1,
+		path:      path,
+		mutex:     new(sync.Mutex),
+		sem:       semaphore.NewWeighted(1),
+	}, nil
+}
+
+// open opens the flock's path and sets the file desriptor. State must be
```
Comment (giuseppe) [inline pkg/flock/flock_unix.go]: typo in `desriptor`

File: pkg/flock/flock_unix.go [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+// synchronized by the caller.
+func (f *flock) open() error {
+	if f.fd != -1 || f.flockType != flockUnlocked {
+		return fmt.Errorf("%v: trying to open improper Flock %q %d %d", errInternal, f.path, f.fd, f.flockType)
```
Comment (giuseppe) [inline pkg/flock/flock_unix.go]: please use `errors.Wrapf`

Comment (vrothberg) [reply pkg/flock/flock_unix.go]: Initially I had that but the problem is that Wrapf mixes the order. I want to have a "internal error ..." but using Wrapf would move that to the end.

File: pkg/flock/flock_windows.go (hunk 1 above)
Comment (vrothberg) [reply pkg/flock/flock_windows.go]: Oh yes, I should. Thanks!

**Comment by vrothberg** [comment]
Marking it as WIP until there's a user of this API, presumably the copy-blob-detection in c/image.

File: pkg/flock/flock_unix.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+	"github.com/pkg/errors"
+	"golang.org/x/sync/semaphore"
+	"golang.org/x/sys/unix"
+)
+
+// New returns a Flock for path.
+func New(path string) (flock, error) {
+	// Make sure we can operate on the path.
+	fd, err := unix.Open(path, os.O_RDWR|os.O_CREATE, unix.S_IRUSR|unix.S_IWUSR)
```
Comment (rhatdan) [inline pkg/flock/flock_unix.go]: Should we use an unix.Access check for this rather then opening and closing?  Would that be faster? And then only create the file if it does not exists.

**Comment by rhatdan** [comment]
@nalind WDYT?

File: pkg/flock/flock.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
+// flock exposes an API around flock(2) but allows for synchronizing within and
+// across process spaces, making it suitable for currently running threads and
```
Comment (vrothberg) [inline pkg/flock/flock.go]: s/currently/concurrently/

**Comment by rhatdan** [comment]
@vrothberg Makes sense, since I envisioned @nalind Blocking this if it was not used in containers/storage.

File: pkg/flock/flock_unix.go (hunk 3 above)
Comment (nalind) [reply pkg/flock/flock_unix.go]: Perhaps `errors.WithStack()`?

**Comment by rhatdan** [comment]
@vrothberg Lets talk about this in the next meeting.

**Comment by vrothberg** [comment]
Ping. @rhatdan, let's check if we still/really need this PR.

**Comment by rhatdan** [comment]
@vrothberg Needs a rebase.  New conversations with @saschagrunert might get us a user for this, yet.

**Comment by vrothberg** [comment]
@rhatdan, I prefer to wait until we're certain there will be a user.

**Comment by rhatdan** [comment]
@vrothberg Since this has been around for 2+ years, can we close it?

**Pull request closed without merging**